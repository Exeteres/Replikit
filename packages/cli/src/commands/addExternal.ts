import { program } from "commander";
import {
    setupConfiguration,
    getProjectManager,
    createExternalProjectName,
    logger,
    validateDirectoryName
} from "@replikit/cli";
import { prompt } from "inquirer";
import { pathExists, ensureDir } from "fs-extra";
import { join } from "path";

interface FirstPromptResult {
    url: string;
    name: string;
}

interface SecondPromptResult {
    modules: string[];
}

interface AdditionalPromptResult {
    postfix: string;
}

const command = program
    .command("add-external")
    .description("Add an external repository with Replikit modules");
command.action(async options => {
    await setupConfiguration(options.config);
    const project = await getProjectManager();
    const firstResult = await prompt<FirstPromptResult>([
        {
            type: "input",
            name: "url",
            message: "Repository url"
        },
        {
            type: "input",
            name: "name",
            message: "Name",
            default: (result: FirstPromptResult) => createExternalProjectName(result.url),
            validate: validateDirectoryName
        }
    ]);
    try {
        await ensureDir(project.externalPath);
        await project.git.addSubmodule(firstResult.url, firstResult.name);
    } catch (e) {
        logger.fatal("Error while adding the external repository", e);
    }
    let path = firstResult.name;
    const modulesPath = join(project.externalPath, firstResult.name, "modules");
    const modulesPathExists = await pathExists(modulesPath);
    if (!modulesPathExists) {
        const additionalResult = await prompt<AdditionalPromptResult>({
            type: "input",
            name: "postfix",
            message: "Relative path to Replikit root directory in the repository",
            validate: async (input: string) => {
                const modulesPath = join(project.externalPath, firstResult.name, input, "modules");
                const modulesPathExists = await pathExists(modulesPath);
                return modulesPathExists || 'Specified directory doesn\'t contain "modules"';
            }
        });
        path = join(firstResult.name, additionalResult.postfix);
    }
    try {
        await project.addExternalRepo(path);
        const allModules = await project.getExternalModuleNames(path);
        logger.info("Repository submodule added");
        const secondResult = await prompt<SecondPromptResult>([
            {
                type: "checkbox",
                name: "modules",
                message: "External modules",
                choices: allModules,
                validate: (input: string[]) => !!input.length || "No modules selected"
            }
        ]);
        await project.addLocalModules(secondResult.modules);
        await project.resolveExternalDependencies(path);
        logger.info("External modules added successfuly");
    } catch (e) {
        logger.fatal("Error while adding the external repository", e);
    }
});
