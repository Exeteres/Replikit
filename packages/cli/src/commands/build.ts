import { program } from "commander";
import {
    loadConfiguration,
    createWebpackConfiguration,
    getProjectManager,
    logger
} from "@replikit/cli";
import { updateConfig, config } from "@replikit/core";
import webpack from "webpack";
import { inspect } from "util";
import { remove, pathExists } from "fs-extra";
import { CliConfiguration } from "@replikit/cli/typings";

const command = program
    .command("build")
    .option("--print-config")
    .option("--development")
    .description("Build all modules to production-ready bundle");
command.action(async options => {
    const newConfig = await loadConfiguration(options.config);
    updateConfig(newConfig);

    const projectManager = await getProjectManager();
    const webpackConfig = createWebpackConfiguration(
        projectManager,
        config.cli as CliConfiguration,
        options.development
    );

    if (options.printConfig) {
        const message = inspect(webpackConfig, {
            showHidden: false,
            depth: null,
            colors: true
        });
        return console.log(message);
    }

    if (await pathExists(webpackConfig.output!.path!)) {
        await remove(webpackConfig.output!.path!);
    }

    webpack(webpackConfig, (error, stats) => {
        if (error) {
            return logger.error("Error while webpack compilation", error);
        }

        const message = stats.toString({
            errors: true,
            colors: true,
            all: false,
            assets: true,
            hash: true,
            version: true,
            builtAt: true
        });
        console.log(message.replace(/\[32m/gm, "[34m"));
    });
});
