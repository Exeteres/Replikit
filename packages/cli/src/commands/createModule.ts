import {
    getProjectManager,
    loadConfiguration,
    validateDirectoryName,
    availableModules,
    createModule,
    logger
} from "@replikit/cli";
import { program } from "commander";
import { prompt } from "inquirer";
import { updateConfig } from "@replikit/core";

const command = program
    .command("create-module")
    .description("Create a new module");
command.action(async options => {
    const config = await loadConfiguration(options.config);
    updateConfig(config);
    const result = await prompt([
        {
            type: "input",
            name: "moduleName",
            message: "Module name",
            validate: validateDirectoryName
        },
        {
            type: "confirm",
            name: "addLogger",
            message: "Do you want to add a logger to the module?"
        },
        {
            type: "checkbox",
            name: "modules",
            message: "Modules",
            choices: availableModules
        }
    ]);
    const manager = await getProjectManager();
    try {
        await createModule(
            manager,
            result.moduleName,
            result.modules,
            result.addLogger
        );
        logger.info("Module created successfuly");
    } catch (e) {
        logger.fatal("Error while creating the module", e);
    }
});
