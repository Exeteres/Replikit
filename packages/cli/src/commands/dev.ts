import { program } from "commander";
import execa from "execa";
import { resolve } from "path";

const command = program
    .command("dev")
    .description("Start in dev mode")
    .option(
        "--inspect-brk <port>",
        "Enable NodeJS debugger with specified port"
    );
command.action(async options => {
    const tsndPath = require.resolve("ts-node-dev/bin/ts-node-dev");
    const workerPath = resolve(__dirname, "../worker/dev.js");
    const args = [
        "--respawn",
        "--clear",
        "--no-notify",
        "--files",
        "--exit-child",
        // "--transpile-only",
        "--watch",
        "replikit.config.ts"
    ];
    if (options.inspectBrk) {
        args.push(`--inspect-brk=${options.inspectBrk}`);
    }
    args.push("--", workerPath);
    if (options.config) {
        args.push(options.config);
    }
    await execa(tsndPath, args, {
        stdin: process.stdin,
        stdout: process.stdout,
        stderr: process.stderr
    });
});
