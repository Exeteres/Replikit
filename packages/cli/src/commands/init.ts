import {
    detectYarn,
    checkFolderIsEmpty,
    validateDirectoryName,
    logger,
    initProject,
    availableStaticModules,
    availableModules,
    createModule
} from "@replikit/cli";
import { prompt } from "inquirer";
import { pathExists } from "fs-extra";
import { resolve } from "path";
import { program } from "commander";

interface PromptResult {
    projectName: string;
    useYarn: boolean;
    useLerna: boolean;
    staticModules: string[];
    createModule: boolean;
    moduleName: string;
    addLogger: boolean;
    modules: string[];
}

const command = program.command("init").description("Create a new project");
command.action(async () => {
    const yarnExists = await detectYarn();
    const cwd = process.cwd();
    const isEmpty = await checkFolderIsEmpty(cwd);
    const result = await prompt<PromptResult>([
        {
            type: "input",
            message: "Project name",
            name: "projectName",
            when: !isEmpty,
            validate: async (path): Promise<true | string> => {
                const message = validateDirectoryName(path);
                if (message !== true) {
                    return message;
                }
                const exists = await pathExists(resolve(cwd, path));
                return !exists || "Item with the same name already exists";
            }
        },
        {
            type: "confirm",
            name: "useYarn",
            default: true,
            message: "Do you want to use Yarn?",
            when: yarnExists
        },
        {
            type: "confirm",
            name: "useLerna",
            default: false,
            message: "Do you want to use Lerna in addition to Yarn?",
            when: (result): boolean => result.useYarn
        },
        {
            type: "checkbox",
            name: "staticModules",
            message: "Static modules",
            choices: availableStaticModules
        },
        {
            type: "confirm",
            name: "createModule",
            message: "Do you want to create a new module?",
            default: true
        },
        {
            type: "input",
            name: "moduleName",
            message: "Module name",
            validate: validateDirectoryName,
            when: (result): boolean => result.createModule
        },
        {
            type: "confirm",
            name: "addLogger",
            message: "Do you want to add a logger to the module?",
            when: (result): boolean => result.createModule
        },
        {
            type: "checkbox",
            name: "modules",
            message: "Modules",
            choices: availableModules,
            when: (result): boolean => result.createModule
        }
    ]);
    const root = isEmpty ? cwd : resolve(cwd, result.projectName);
    const useLerna = !result.useYarn || result.useLerna;
    try {
        const manager = await initProject(
            root,
            useLerna,
            result.useYarn,
            result.staticModules
        );
        if (result.createModule) {
            await createModule(
                manager,
                result.moduleName,
                result.modules,
                result.addLogger
            );
        }
        logger.info("Project created successfuly");
    } catch (e) {
        logger.fatal("Error while creating the project", e);
    }
});
