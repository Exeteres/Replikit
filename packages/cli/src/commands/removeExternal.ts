import {
    getProjectManager,
    logger,
    promptDeleteConfirmation,
    setupConfiguration
} from "@replikit/cli";
import { program } from "commander";

const command = program
    .command("remove-external")
    .arguments("<repository>")
    .description("Remove external repository from the project")
    .option("-y, --yes", "Skip confirmation");
command.action(async (repository, options) => {
    await setupConfiguration(options.config);
    const confirmed = await promptDeleteConfirmation(options.yes);
    if (!confirmed) {
        return;
    }
    const project = await getProjectManager();
    await project.removeExternalRepository(repository);
    logger.info("External repository removed successfully");
});
