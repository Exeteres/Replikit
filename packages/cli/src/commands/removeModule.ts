import {
    getProjectManager,
    logger,
    promptDeleteConfirmation,
    setupConfiguration
} from "@replikit/cli";
import { program } from "commander";

const command = program
    .command("remove-module")
    .arguments("<module>")
    .description("Remove module from the project")
    .option("-y, --yes", "Skip confirmation");
command.action(async (module, options) => {
    await setupConfiguration(options.config);
    const confirmed = await promptDeleteConfirmation(options.yes);
    if (!confirmed) {
        return;
    }
    const project = await getProjectManager();
    const removed = await project.removeModule(module);
    if (removed) {
        logger.info("Module removed successfully");
    } else {
        logger.error("Module not found");
    }
});
