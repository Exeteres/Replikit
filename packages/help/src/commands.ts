import { command, commands } from "@replikit/commands";
import { descriptions } from "@replikit/help";
import { fromCode } from "@replikit/messages";

command("help")
    .handler(context => {
        const message = descriptions.render(
            commands.getCommands(),
            context.account.language
        );
        return fromCode(message);
    })
    .register();
