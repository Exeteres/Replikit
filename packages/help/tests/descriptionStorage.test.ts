import {
    DescriptionStorage,
    InvalidCommandDescriptionError,
    DefaultLocaleNotFoundError
} from "@replikit/help";
import { CommandStorage } from "@replikit/commands";
import { command } from "@replikit/commands";

describe("DescriptionStorage", () => {
    it("should add a locale with description messages", () => {
        const descriptionStorage = new DescriptionStorage();
        descriptionStorage.add("en", { help: "Show help" });
        descriptionStorage.add("en", { test: "Test" });

        expect(descriptionStorage).toMatchSnapshot();
    });

    it("should render a help message with multiple commands", () => {
        const descriptionStorage = new DescriptionStorage();
        descriptionStorage.add("en", {
            test: "Test",
            calc: { sum: "Sum two numbers", diff: "Diff two numbers" }
        });

        const commands = new CommandStorage();
        commands.register(
            command("test")
                .handler(() => void 0)
                .build()
        );
        commands.register(
            command("calc")
                .commands(
                    command("sum")
                        .required("a", Number)
                        .required("b", Number)
                        .handler(() => void 0),
                    command("diff")
                        .required("a", Number)
                        .required("b", Number)
                        .handler(() => void 0)
                )
                .build()
        );

        const message = descriptionStorage.render(commands.getCommands(), "en");
        expect(message).toMatchSnapshot();
    });

    it("should render a help message with overloaded command", () => {
        const descriptionStorage = new DescriptionStorage();
        descriptionStorage.add("en", { test: "Test" });

        const commands = new CommandStorage();
        commands.register(
            command("test")
                .handler(() => void 0)
                .build()
        );
        commands.register(
            command("test")
                .required("test", Number)
                .handler(() => void 0)
                .build()
        );

        const message = descriptionStorage.render(commands.getCommands(), "en");
        expect(message).toMatchSnapshot();
    });

    it("should throw an error when render an invalid structure 1", () => {
        const descriptionStorage = new DescriptionStorage();
        descriptionStorage.add("en", {
            test: {}
        });

        const commands = new CommandStorage();
        commands.register(
            command("test")
                .handler(() => void 0)
                .build()
        );

        const action = () =>
            descriptionStorage.render(commands.getCommands(), "en");
        expect(action).toThrow(InvalidCommandDescriptionError);
    });

    it("should throw an error when render an invalid structure 2", () => {
        const descriptionStorage = new DescriptionStorage();
        descriptionStorage.add("en", {
            calc: "Calc"
        });

        const commands = new CommandStorage();
        commands.register(
            command("calc")
                .commands(
                    command("sum")
                        .required("a", Number)
                        .required("b", Number)
                        .handler(() => void 0),
                    command("diff")
                        .required("a", Number)
                        .required("b", Number)
                        .handler(() => void 0)
                )
                .build()
        );

        const action = () =>
            descriptionStorage.render(commands.getCommands(), "en");
        expect(action).toThrow(InvalidCommandDescriptionError);
    });

    it("should throw an error when default locale not exists", () => {
        const descriptionStorage = new DescriptionStorage();
        const action = () => descriptionStorage.render([], "en");
        expect(action).toThrow(DefaultLocaleNotFoundError);
    });
});
