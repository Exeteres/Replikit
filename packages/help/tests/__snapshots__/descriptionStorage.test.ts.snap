// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DescriptionStorage should add a locale with description messages 1`] = `
DescriptionStorage {
  "descriptionMap": Map {
    "en" => Object {
      "help": "Show help",
      "test": "Test",
    },
  },
}
`;

exports[`DescriptionStorage should render a help message with multiple commands 1`] = `
"# Test
/test

# Sum two numbers
/calc sum {a} {b}

# Diff two numbers
/calc diff {a} {b}
"
`;

exports[`DescriptionStorage should render a help message with overloaded command 1`] = `
"# Test
/test
/test {test}"
`;
